src/ui/public/visualize/visualization.js

```diff
@@ -0,0 +1,147 @@
+    link: function ($scope, $el) {
+      const minVisChartHeight = 180;
+      const resizeChecker = new ResizeChecker($el);
+
+      $scope.showSpyPanel = $scope.vis && $scope.vis.showSpyPanel || false;
+
+
+      //todo: lets make this a simple function call.
+      const getVisEl = jQueryGetter('.visualize-chart');
+      const getVisContainer = jQueryGetter('.vis-container');
+      const getSpyContainer = jQueryGetter('.visualize-spy-container');
+
+      // Show no results message when isZeroHits is true and it requires search
+      $scope.showNoResultsMessage = function () {
+        const requiresSearch = _.get($scope, 'vis.type.requiresSearch');
+        const isZeroHits = _.get($scope,'visData.hits.total') === 0;
+        const shouldShowMessage = !_.get($scope, 'vis.params.handleNoResults');
+
+        return Boolean(requiresSearch && isZeroHits && shouldShowMessage);
+      };
+
+      const legendPositionToVisContainerClassMap = {
+        top: 'vis-container--legend-top',
+        bottom: 'vis-container--legend-bottom',
+        left: 'vis-container--legend-left',
+        right: 'vis-container--legend-right',
+      };
+
+      $scope.getVisContainerClasses = function () {
+        return legendPositionToVisContainerClassMap[$scope.vis.params.legendPosition];
+      };
+
+      $scope.spy = {};
+      $scope.spy.mode = ($scope.uiState) ? $scope.uiState.get('spy.mode', {}) : {};
+
+      const applyClassNames = function () {
+        const $visEl = getVisContainer();
+        const $spyEl = getSpyContainer();
+        if (!$spyEl) return;
+
+        const fullSpy = ($scope.spy.mode && ($scope.spy.mode.fill || $scope.fullScreenSpy));
+
+        $visEl.toggleClass('spy-only', Boolean(fullSpy));
+        $spyEl.toggleClass('only', Boolean(fullSpy));
+
+        $timeout(function () {
+          if (shouldHaveFullSpy()) {
+            $visEl.addClass('spy-only');
+            $spyEl.addClass('only');
+          }
+        }, 0);
+      };
+
+      const loadingDelay = config.get('visualization:loadingDelay');
+      $scope.loadingStyle = {
+        '-webkit-transition-delay': loadingDelay,
+        'transition-delay': loadingDelay
+      };
+
+      function shouldHaveFullSpy() {
+        const $visEl = getVisEl();
+        if (!$visEl) return;
+
+        return ($visEl.height() < minVisChartHeight)
+          && _.get($scope.spy, 'mode.fill')
+          && _.get($scope.spy, 'mode.name');
+      }
+
+      // spy watchers
+      $scope.$watch('fullScreenSpy', applyClassNames);
+
+      $scope.$watchCollection('spy.mode', function () {
+        $scope.fullScreenSpy = shouldHaveFullSpy();
+        applyClassNames();
+      });
+
+      const Visualization = $scope.vis.type.visualization;
+
+      //todo: make this not a jquery element
+      const visualization = new Visualization(getVisEl()[0], $scope.vis);
+
+      const renderFunction = _.debounce(() => {
+        visualization.render($scope.visData)
```

Comment:
i think there is no need for extra parameters saying whats changed, as it will be different from vis to vis in what its interested in. some vis might want to know if params changed, some might want to check specific parameters, some might want to check specific aggs .. so its up to visualization to implement this check.

in the long run we are moving towards react where this kind of checking probably won't be necessary at all.

regarding the second suggestion to pass params with the render method ... the params are already available on the vis object, but i can still pass them in here as well.